GitUI/Theming/LocalHook.cs

```diff
@@ -0,0 +1,480 @@
+        /// If you know how to write such native delegates you won't need internet resources of course.
+        /// I recommend using C++.NET which allows you to just copy the related windows API to your managed
+        /// class and thread it as delegate without any changes. This will also speed up the whole thing
+        /// because no unnecessary marshalling is required! C++.NET is also better in most cases because you
+        /// may access the whole native windows API from managed code without any effort what significantly eases
+        /// writing of hook handlers.
+        /// </para>
+        /// <para>
+        /// The given delegate is automatically prevented from being garbage collected until the hook itself
+        /// is collected...
+        /// </para>
+        /// </remarks>
+        /// <param name="InTargetProc">A target entry point that should be hooked.</param>
+        /// <param name="InNewProc">A handler with the same signature as the original entry point
+        /// that will be invoked for every call that has passed the Fiber Deadlock Barrier and various integrity checks.</param>
+        /// <returns>
+        /// A handle to the newly created hook.
+        /// </returns>
+        /// <exception cref="OutOfMemoryException">
+        /// Not enough memory available to complete the operation. On 64-Bit this may also indicate
+        /// that no memory can be allocated within a 31-Bit boundary around the given entry point.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// The given function pointer does not map to executable memory (valid machine code) or
+        /// you passed <c>null</c> as delegate.
+        /// </exception>
+        /// <exception cref="NotSupportedException">
+        /// The given entry point contains machine code that can not be hooked.
+        /// </exception>
+        /// <exception cref="InsufficientMemoryException">
+        /// The maximum amount of hooks has been installed. This is currently set to MAX_HOOK_COUNT (1024).
+        /// </exception>
+        public static LocalHook Create(
+            IntPtr InTargetProc,
+            Delegate InNewProc,
+            Object InCallback)
+        {
+            LocalHook Result = new LocalHook();
+
+            Result.m_Callback = InCallback;
+            Result.m_HookProc = InNewProc;
+            Result.m_Handle = Marshal.AllocCoTaskMem(IntPtr.Size);
+            Result.m_SelfHandle = GCHandle.Alloc(Result, GCHandleType.Weak);
+
+            Marshal.WriteIntPtr(Result.m_Handle, IntPtr.Zero);
+
+            try
+            {
+                ////NativeAPI.LhInstallHook(
+                ////    InTargetProc,
+                ////    Marshal.GetFunctionPointerForDelegate(Result.m_HookProc),
+                ////    GCHandle.ToIntPtr(Result.m_SelfHandle),
+                ////    Result.m_Handle);
```

Comment:
All commented code in this file should be updated to this pattern:

```suggestion
#if SUPPORT_THEMES
                NativeAPI.LhInstallHook(
                    InTargetProc,
                    Marshal.GetFunctionPointerForDelegate(Result.m_HookProc),
                    GCHandle.ToIntPtr(Result.m_SelfHandle),
                    Result.m_Handle);
#endif
```